### PR TITLE
Api and files

### LINKED ISSUE #975: aptly CLI should be able to work while aptly web is running
At the moment, one has to stop the aptly web server in order to be able to use the `aptly` command from the terminal. This is because leveldb prevents multiple processes from accessing the same database.

We should lift this limitation. One obvious way would be for the CLI to make use of the API (although this would require the API to be running when using the CLI, which is not always wanted/possible).

### LINKED ISSUE #1017: Serve api and published packages simultaneously
Not sure about the reason making it two separate commands, but the following patch allows to perform both in one call.

```patch
diff --git a/api/router.go b/api/router.go
index fdfbb94..e164d97 100644
--- a/api/router.go
+++ b/api/router.go
@@ -1,6 +1,7 @@
 package api
 
 import (
+ "github.com/aptly-dev/aptly/aptly"
        "net/http"
 
        ctx "github.com/aptly-dev/aptly/context"
@@ -15,6 +16,8 @@ func Router(c *ctx.AptlyContext) http.Handler {
 
        router := gin.Default()
        router.Use(gin.ErrorLogger())
+ publicPath := context.GetPublishedStorage("").(aptly.FileSystemPublishedStorage).PublicPath()
+ router.Static("/repos/", publicPath)
 
        if context.Flags().Lookup("no-lock").Value.Get().(bool) {
                // We use a goroutine to count the number of
```

Changes are pretty minimalistic and I'd be glad to see them included.
If there are no objections, I'm ready to make a PR. (disclamer: I'm not a gopher ;) )

### PR DESCRIPTION
**Pull request by mmianl**
Duplicate of #1116 but with system tests.

Fixes #1017, maybe also fixes #975 but not sure about that

Related documentation change: https://github.com/aptly-dev/www.aptly.info/pull/103

## Requirements

All new code should be covered with tests, documentation should be updated. CI should pass.

## Checklist

- [ ] unit-test added (if change is algorithm)
- [x] functional test added/updated (if change is functional)
- [ ] man page updated (if applicable)
- [ ] bash completion updated (if applicable)
- [x] documentation updated
- [ ] author name in `AUTHORS`

### CODE REVIEW

**Comment by aol-nnov** [comment]
image

Seems, I can make some fuss in the mainline. Should I? :-D

**Comment by mmianl** [comment]
I think this should be approved by @randombenj ;)

**Comment by aol-nnov** [comment]
Definitely! )) But technically, for some reason I have the approve button too here and it reads "Merging can be performed automatically with 1 approving review.", soooo.... ))))

**Comment by mmianl** [comment]
This is probably because I cherry-picked your commit to add system tests, so I assume that this makes you a code owner in the scope of this PR.

I might be wrong though...

File: system/t12_api/publish.py [hunk 1]
```diff
@@ -3,6 +3,11 @@
 
 from api_lib import TASK_SUCCEEDED, APITest
 
+try:
+    import requests
+except ImportError:
+    requests = None
+
```
Comment (randombenj) [inline system/t12_api/publish.py:10]: Why is this needed? I think this should be removed

File: api/router.go [hunk 2]
```diff
@@ -51,6 +51,21 @@ func Router(c *ctx.AptlyContext) http.Handler {
 		MetricsCollectorRegistrar.Register(router)
 	}
 
+	if c.Config().ServeInAPIMode {
```
Comment (randombenj) [inline api/router.go:54]: Cool stuff! What do you think if the `/repos/` endpoint would return  a list (as json?) of all 
the published repos?

**Review by randombenj — APPROVED** [review]

File: api/router.go (hunk 2 above)
Comment (aol-nnov) [reply api/router.go:54]: I think, if we really want to extend changes scope to that point, it should be not a json response, but a "directory listing"-like html response similar to that generated by `c.FileFromFS` when requested path is a directory.

File: system/t12_api/publish.py (hunk 1 above)
Comment (mmianl) [reply system/t12_api/publish.py:10]: You are right, I forgot to remove this...

File: api/router.go (hunk 2 above)
Comment (mmianl) [reply api/router.go:54]: How would I determine which repos have been published?

Comment (aol-nnov) [reply api/router.go:54]: @mmianl something like that: https://github.com/aptly-dev/aptly/blob/77e02bf7a363c78f9c82792a569c8a5a496c1f06/cmd/publish_list.go#L85-L96

But we need to filter out only `filesystem` publishes, if I get the idea right.

Comment (randombenj) [reply api/router.go:54]: Agreed with @aol-nnov , directory listing response like would be nice!
But @mmianl if you think this is out of scope I am happy to merge as is/

Comment (aol-nnov) [reply api/router.go:54]: @randombenj I could fiddle with it in a separate PR if my time permits and/or @mmianl won't come with a solution in this PR.

Anyway, current changeset provides viable support for simultaneous serving of files and API and could be merged as is.

**Comment by aol-nnov** [comment]
One more thing came to my mind: it would be nice to align http repo paths during `serve` and `api serve`.

Currently former serves from the root and latter serves from the `/repos`

File: api/router.go (hunk 2 above)
Comment (mmianl) [reply api/router.go:54]: @mmianl @aol-nnov could you provide an example output for this?

Comment (aol-nnov) [reply api/router.go:54]: @mmianl I've left my laptop in the office. I'll try to arrange some example tomorrow

Comment (aol-nnov) [reply api/router.go:55]: @mmianl I've slept with the idea and actually, it is much easier, than listing publishes an filtering them.

All we need is to list publish prefixes. Rest is handled by the `/repos/:storage/*pkgPath` route!

If you grant me write permissions for this PR, I could push a changeset. Meanwhile find a patch for you latest changes:


```patch
diff --git a/api/repos.go b/api/repos.go
index 8a38dbda..11eddb9e 100644
--- a/api/repos.go
+++ b/api/repos.go
@@ -15,30 +15,20 @@ import (
        "github.com/aptly-dev/aptly/task"
        "github.com/aptly-dev/aptly/utils"
        "github.com/gin-gonic/gin"
-       "github.com/rs/zerolog/log"
 )
 
 // GET /repos
-// todo: limit to filesystem repos only
-// todo: change output format
-func reposListInAPIMode(c *gin.Context) {
-       repos := []*deb.PublishedRepo{}
+func reposListInAPIMode(localRepos map[string]utils.FileSystemPublishRoot) gin.HandlerFunc {
 
-       _ = context.NewCollectionFactory().PublishedRepoCollection().ForEach(func(repo *deb.PublishedRepo) error {
-               err := context.NewCollectionFactory().PublishedRepoCollection().LoadComplete(repo, context.NewCollectionFactory())
-               if err != nil {
-                       msg := fmt.Sprintf(
-                               "Error %s found while getting list of published repos (prefix:%s / distribution:%s / component:%s\n).",
-                               err, repo.StoragePrefix(), repo.Distribution, repo.Components())
-                       log.Debug().Msg(msg)
-                       return err
+       return func(c *gin.Context) {
+               res := c.Writer
+               res.WriteString("<html><body><pre>")
+               for publishPrefix := range localRepos {
+                       res.WriteString(fmt.Sprintf("<a href=\"%[1]s/\">%[1]s</a>\n", publishPrefix))
                }
-
-               repos = append(repos, repo)
-               return nil
-       })
-
-       c.JSON(200, repos)
+               res.WriteString("</pre></body></html>")
+               res.Flush()
+       }
 }
 
 // GET /repos/:storage/*pkgPath
diff --git a/api/router.go b/api/router.go
index 1e368192..ea1dc27f 100644
--- a/api/router.go
+++ b/api/router.go
@@ -53,7 +53,7 @@ func Router(c *ctx.AptlyContext) http.Handler {
        }
 
        if c.Config().ServeInAPIMode {
-               router.GET("/repos/", reposListInAPIMode)
+               router.GET("/repos/", reposListInAPIMode(c.Config().FileSystemPublishRoots))
                router.GET("/repos/:storage/*pkgPath", reposServeInAPIMode)
        }
```

Comment (mmianl) [reply api/router.go:55]: Updated accordoing to @aol-nnov's feedback. I changed a few details to make the output of `repos` consistent with the `FileFromFS()` output. I also updated the handler for `/repos/:storage/*pkgPath` according to the above suggestion.

@randombenj, @aol-nnov  what do you think?

File: api/router.go [hunk 3]
```diff
@@ -51,6 +52,12 @@ func Router(c *ctx.AptlyContext) http.Handler {
...
```
Comment (aol-nnov) [inline api/router.go:55]: @mmianl neat!

I have a question, though... `c.Config().ServeInAPIMode` on line 55 that you introduced earlier and `context.Config().FileSystemPublishRoots` on line 56 are, actually from the same context. I'd rather use the same variable for tha sake of consistency. (There are more of such cases down the file, I know, but they are out of scope)

Other than that, LGTM!

Comment (mmianl) [reply api/router.go:55]: Fair enough

**Comment by mmianl** [comment]
@randombenj feel free to merge this anytime now

File: api/router.go (hunk 2 above)
Comment (randombenj) [reply api/router.go:55]: Love it thanks :tada:

**Comment by randombenj** [comment]
Thank you so much for this change @mmianl and @aol-nnov !
